### PR TITLE
[KW-search] Indices on parents array

### LINKED ISSUE #9251: [KW-search] Allow querying children of a given parent via index
See https://dust4ai.slack.com/archives/C050SM8NSPK/p1733837120557819

### PR DESCRIPTION
**Pull request by philipperolet**
Description
---
Fixes https://github.com/dust-tt/dust/issues/9251

The PR will be updated with explain analyze queries to show the indices perform as expected. If they do not, they will be rolled back.

Related important context: https://dust4ai.slack.com/archives/C050SM8NSPK/p1733837120557819

Risks
---
Creating indices could slow down production. Mitigated by 1. running at low peak (lunch or evening) 2. using [CONCURRENTLY](https://dust.tt/w/0ec9852c2f/assistant/59NYjdiRr3):

CREATE INDEX CONCURRENTLY is designed to be low-impact:
- Doesn't block writes/reads on the table
- Takes ~2-3x longer than regular CREATE INDEX
- Can be cancelled safely with pg_cancel_backend() or pg_terminate_backend()

Main impact is on disk I/O and some CPU

Two caveats:
- Requires ~2x disk space during creation
- Can't run inside a transaction block

Deploy
---
- run index creation on prodbox
- post explain analyze queries here
- either rollback, or deploy core

### CODE REVIEW

File: core/src/stores/migrations/202412011_indices_parents.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+CREATE INDEX CONCURRENTLY idx_parents_second ON data_sources_nodes (data_source, (parents[2]));
+CREATE INDEX CONCURRENTLY idx_parents_single ON data_sources_nodes (data_source, (array_length(parents, 1) = 1));
```
Comment (tdraier) [inline core/src/stores/migrations/202412011_indices_parents.sql:2]: include the name of the table in the index : 
```suggestion
CREATE INDEX CONCURRENTLY idx_data_sources_nodes_parents_second ON data_sources_nodes (data_source, (parents[2]));
CREATE INDEX CONCURRENTLY idx_data_sources_nodes_parents_single ON data_sources_nodes (data_source, (array_length(parents, 1) = 1));
```

File: core/src/stores/store.rs [hunk 2]
```diff
@@ -675,6 +675,10 @@ pub const SQL_INDEXES: [&'static str; 33] = [
         idx_data_sources_nodes_table ON data_sources_nodes(\"table\");",
     "CREATE INDEX IF NOT EXISTS
         idx_data_sources_nodes_folder ON data_sources_nodes(folder);",
+    "CREATE INDEX IF NOT EXISTS
+        idx_parents_second ON data_sources_nodes (data_source, (parents[2]));",
+    "CREATE INDEX IF NOT EXISTS
+        idx_parents_single ON data_sources_nodes (data_source, (array_length(parents, 1) = 1));",
```
Comment (tdraier) [inline core/src/stores/store.rs:681]:
```suggestion
    "CREATE INDEX IF NOT EXISTS
        idx_data_sources_nodes_parents_second ON data_sources_nodes (data_source, (parents[2]));",
    "CREATE INDEX IF NOT EXISTS
        idx_data_sources_nodes_parents_single ON data_sources_nodes (data_source, (array_length(parents, 1) = 1));",
```

**Review by tdraier — APPROVED** [review]

File: core/src/stores/migrations/20241211_indices_parents.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+CREATE INDEX CONCURRENTLY idx_data_sources_nodes_parents_second ON data_sources_nodes (data_source, (parents[2]));
```
Comment (spolu) [inline core/src/stores/migrations/20241211_indices_parents.sql:1]: Why parents[2] ?

Comment (philipperolet) [reply core/src/stores/migrations/20241211_indices_parents.sql:1]: in postgres arrays start at 1

**Comment by philipperolet** [comment]
327 => our Notion
parent id used => the Engineering page
**Conclusion: index works for parents[2] case**
```
dust-api=> EXPLAIN ANALYZE SELECT * FROM data_sources_nodes WHERE data_source = 327 AND parents[2] = 'e0f834b5-be5a-4356-9baa-f76e9c41adf2';
                                                                          QUERY PLAN                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------
 Index Scan using idx_data_sources_nodes_parents_second on data_sources_nodes  (cost=0.56..47.88 rows=11 width=358) (actual time=0.648..0.710 rows=74 loops=1)
   Index Cond: ((data_source = 327) AND (parents[2] = 'e0f834b5-be5a-4356-9baa-f76e9c41adf2'::text))
 Planning Time: 0.603 ms
 Execution Time: 0.771 ms
(4 rows)
```

**Comment by philipperolet** [comment]
327 => our notion still
**Conclusion: index works for single-parent array case**
```
dust-api=> EXPLAIN ANALYZE SELECT * FROM data_sources_nodes WHERE data_source = 327 AND array_length(parents, 1) = 1;
                                                                            QUERY PLAN                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Index Scan using idx_data_sources_nodes_parents_single on data_sources_nodes  (cost=0.43..4462.94 rows=11 width=358) (actual time=0.022..0.182 rows=243 loops=1)
   Index Cond: ((data_source = 327) AND ((array_length(parents, 1) = 1) = true))
 Planning Time: 0.945 ms
 Execution Time: 0.243 ms
(4 rows)
```

**Pull request merged**